drivers/mspi/mspi_nrfe.c

```diff
@@ -0,0 +1,718 @@
+			  const enum mspi_dev_cfg_mask param_mask, const struct mspi_dev_cfg *cfg)
+{
+	const struct mspi_nrfe_config *drv_cfg = dev->config;
+	struct mspi_nrfe_data *drv_data = dev->data;
+	int rc;
+
+	if (param_mask & MSPI_DEVICE_CONFIG_MEM_BOUND) {
+		if (cfg->mem_boundary) {
+			LOG_ERR("Auto CE break is not supported.");
+			return -ENOTSUP;
+		}
+	}
+
+	if (param_mask & MSPI_DEVICE_CONFIG_BREAK_TIME) {
+		if (cfg->time_to_break) {
+			LOG_ERR("Auto CE break is not supported.");
+			return -ENOTSUP;
+		}
+	}
+
+	if (param_mask & MSPI_DEVICE_CONFIG_FREQUENCY) {
+		if (cfg->freq > drv_cfg->mspicfg.max_freq) {
+			LOG_ERR("Invalid frequency: %u, MAX: %u", cfg->freq,
+				drv_cfg->mspicfg.max_freq);
+			return -EINVAL;
+		}
+	}
+
+	if (param_mask & MSPI_DEVICE_CONFIG_IO_MODE) {
+		rc = check_io_mode(cfg->io_mode);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	if (param_mask & MSPI_DEVICE_CONFIG_DATA_RATE) {
+		if (cfg->data_rate != MSPI_DATA_RATE_SINGLE) {
+			LOG_ERR("Only single data rate is supported.");
+			return -ENOTSUP;
+		}
+	}
+
+	if (param_mask & MSPI_DEVICE_CONFIG_DQS) {
+		if (cfg->dqs_enable) {
+			LOG_ERR("DQS signal is not supported.");
+			return -ENOTSUP;
+		}
+	}
+
+	memcpy((void *)&drv_data->dev_cfg, (void *)cfg, sizeof(drv_data->dev_cfg));
+	drv_data->dev_id = *dev_id;
+
+	return send_config(NRFE_MSPI_CONFIG_DEV, (void *)cfg, sizeof(struct mspi_dev_cfg));
+}
+
+static int api_get_channel_status(const struct device *dev, uint8_t ch)
+{
+	return 0;
+}
+
+/**
+ * @brief Send a transfer packet to the eMSPI controller.
+ *
+ * @param dev eMSPI controller device
+ * @param packet Transfer packet containing the data to be transferred
+ * @param timeout Timeout in milliseconds
+ *
+ * @retval 0 on success
+ * @retval -ENOTSUP if the packet is not supported
+ * @retval -ENOMEM if there is no space in the buffer
+ * @retval -ETIMEDOUT if the transfer timed out
+ */
+static int xfer_packet(struct mspi_xfer_packet *packet, uint32_t timeout)
+{
+	int rc;
+	uint32_t struct_size = sizeof(struct mspi_xfer_packet);
+	uint32_t len = struct_size + packet->num_bytes + 1;
+	uint8_t buffer[len];
+	enum nrfe_mspi_opcode opcode = (packet->dir == MSPI_RX) ? NRFE_MSPI_TXRX : NRFE_MSPI_TX;
+
+	buffer[0] = (uint8_t)opcode;
+	memcpy((void *)&buffer[1], (void *)packet, struct_size);
+	memcpy((void *)(&buffer[1] + struct_size), (void *)packet->data_buf, packet->num_bytes);
+
+	rc = mspi_ipc_data_send(opcode, buffer, len);
+	if (rc < 0) {
+		LOG_ERR("%d: Packet transfer error: %d", __LINE__, rc);
+	}
+
+	rc = nrfe_mspi_wait_for_response(opcode, timeout);
+	if (rc < 0) {
+		LOG_ERR("%d: FLPR Xfer response timeout: %d", __LINE__, rc);
+		return rc;
+	}
+
+	/* Wait for the transfer to complete and receive data. */
+	if (packet->dir == MSPI_RX) {
+		if ((ipc_receive_buffer != NULL) && (ipc_received > 0)) {
+			memcpy((void *)packet->data_buf, (void *)ipc_receive_buffer, ipc_received);
+			packet->num_bytes = ipc_received;
```

Comment:
Shouldn't that be the other way around?:
```suggestion
			memcpy((void *)ipc_receive_buffer, (void *)packet->data_buf, ipc_received);
			packet->num_bytes = ipc_received;
```
Or maybe these two lines should be deleted since copying is supposed to be done in `ipc_recv_clbk`?

Reply:
>Shouldn't that be the other way around?:

This is for RX path, so it should be copied from ipc buffers to the packet.

>Or maybe these two lines should be deleted since copying is supposed to be done in ipc_recv_clbk?

I think it should stay, copying should not be done in interrupt context.

Reply:
Okay, I can see now that it is probably done to return the received data to the user. I guess there is not much that we can do about it due to the Zephyr API. Could there at least be a comment explaining why received data is copied to the `data_buf` in the packet? I was convinced that `data_buf` can only store TX data, so this line confused me a lot when I first read it.

Reply:
I'll add a comment in a separate PR to not delay this one.